src/miniscript/astelem.rs

```diff
@@ -815,3 +815,45 @@ impl<Pk: MiniscriptKey, Ctx: ScriptContext> Terminal<Pk, Ctx> {
         }
     }
 }
+
+impl<'a, Pk: MiniscriptKey, Ctx: ScriptContext> IntoIterator for &'a Terminal<Pk, Ctx> {
+    type Item = &'a Pk;
+    type IntoIter = Box<dyn Iterator<Item = &'a Pk> + 'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        use std::iter;
+
+        match self {
+            Terminal::PkK(ref pk) => Box::new(iter::once(pk)),
+            Terminal::True
+            | Terminal::False
+            | Terminal::PkH(_)
```

Comment:
Afaik there exists the quite common case that the key "hash" is actually also a key in our representation of miniscript (e.g. when using a descriptor to derive child descriptors). For the key mapping functions there exist very ugly-named different functions for that (`TranslatePk1`-`TranslatePk3`). It might be worth thinking about this here too, because a user may not care about if it is a p2pkh or p2pk node and just wants to treat them all the same.

Reply:
Hmm, I think only for `TranslatePk2` that could be relevant here. Though if the "pubkey hash" type is secretly a key type, that's something pretty messy. It seems that indeed for `DescriptorPublicKey` that is the case and the `Self::Hash` type is set to `Self`. That's pretty neat.

I wonder if we can do some conditional on the `K::Hash` type also being K or something.

Reply:
Hmm, it would create conflicting implementations and also it's not possible to do equality conditions in where clauses. So I can't see how it can be done without creating messy alternative traits and the thing I liked about this impl is that it doesn't need a trait.

As an alternative, I thought about having an IntoIterator where the output iter type is an iterator of Pk::Hash items and we'd call hash on all keys, so that you'd get all the key hashes. But I think if Pk and Pk::Hash are identical, it might again cause problems of duplicate implementations. Because then the type system won't figure out which iterator you want.

Reply:
We could yield an iterator of `miniscript::ForEach` elements. That name is a bit unfortunately chosen though because it's named with the use case it's used in instead of actually naming what it is.

Reply:
See my comment here: https://github.com/rust-bitcoin/rust-miniscript/pull/252#issuecomment-900497210